Fixing S.S.Cryptography.Xml tests

krwq opened the pull request:
cc @anthonylangsworth @tintoy @peterwurzinger

danmosemsft commented:
I asked the vendors to make a sweep over the weekend on corefx to change eg
`[Fact(Skip = "https://github.com/dotnet/corefx/issues/16685")]`
into
```
[Fact]
[ActiveIssue(16685)]
```

krwq commented:
@danmosemsft I have intentionally not marked it as ActiveIssue becuase of https://github.com/dotnet/corefx/issues/16686 which is pretty annoying when working with VS

krwq commented:
@tintoy @anthonylangsworth - I've just made a fix here for some of the yesteday's disabled test

danmosemsft commented:
Oh right..

krwq commented:
or not - something is failing on non Windows - I'll take a look later

tintoy commented:
Sorry @krwq - would love to help out but today's looking like a wash-out, I'm nowhere near my dev machine :-/

Happy to help out again tomorrow though!

(it's midday here)

krwq commented:
It looks like just escaping issue. Not sure how do you properly escape entity references
```
System.Xml.XmlException : An XML comment cannot contain '--', and '-' cannot be the last character. Line 12, position 38.
```
The path seems to be containing three dashes in a row

tintoy commented:
I think you can replace it with `&#x002D;`...

tintoy commented:
Sorry if I was at my dev machine I could try it out but I'm on a low-powered notebook at the moment.

krwq commented:
wait - I'm not sure why I used 169 instead of 2D - probably copy paste fail

tintoy commented:
Yeah I sometimes get it confused with en-dash (the long dash) but that's a bigger value I believe.

krwq commented:
```
        System.Xml.XmlException : Fragment identifier '#2D;&#2D;&#2D;4526f5ff\bin\AnyOS.AnyCPU.Debug\System.Security.Cryptography.Xml.Tests\netcoreapp\XmlDsigExcC14NTransformTest.ExcC14NSpecExample5.txt' cannot be part of the system identifier 'D:\j\workspace\windows_nt_de&#2D;&#2D;&#2D;4526f5ff\bin\AnyOS.AnyCPU.Debug\System.Security.Cryptography.Xml.Tests\netcoreapp\XmlDsigExcC14NTransformTest.ExcC14NSpecExample5.txt'. Line 4, position 23.
```
we need some different way of escaping I guess

tintoy commented:
No worries - I may try expressing it as a `file://` URI to see if that makes a difference.

tintoy commented:
Ah, saw your other reply - yeah I'd rather just get rid of the physical files!

krwq commented:
@tintoy URI is worth trying before making resolver - do you want to pick up from here? Btw do you have access to CI logs?

tintoy commented:
Yep, I see the CI results - it's open to collaborators as well as org members from memory...

tintoy commented:
BTW, yes, I'll take a look sometime this weekend - either URI or `XmlResolver` should have them working by the end of the weekend.

krwq commented:
Thank you @tintoy

anthonylangsworth commented:
@krwq @tintoy I have (hopefully) resolved the concurrency issues with https://github.com/krwq/corefx/pull/1. If I have pushed to the wrong branch, please let me know. There appear to be a few related branches and it is hard to keep up.

krwq commented:
@anthonylangsworth - you're right, I need to clean up my branches on my fork :wink: Thank you for help!

krwq commented:
@anthonylangsworth still some tiny failures, likely some slash vs backslash issue although I haven't investigated

krwq commented:
@anthonylangsworth Thank you again!

krwq commented:
@anthonylangsworth I'll be taking over from where you've left now

krwq commented:
@anthonylangsworth - Nice, everything green! Good job!!!

krwq commented:
@anthonylangsworth @tintoy seems to me like dsaKeyValue.Key.ExportParameters(false); returns object whose Seed is null on linux - I'll investigate it

bartonjs commented on src/System.Security.Cryptography.Xml/src/System/Security/Cryptography/Xml/Transform.cs:
```diff
@@ -204,7 +204,7 @@ public Hashtable PropagatedNamespaces
                 if (reference != null)
                     namespaces = reference._namespaces;
-                else if (signedXml._context != null)
+                else if (signedXml?._context != null)
```
netfx doesn't `?.` here.  Is this a netfx bug, or a test bug, that it needs to change?

krwq replied:
@anthonylangsworth, do you remember why it needed to change?

krwq replied:
@anthonylangsworth - I'm looking at the code and possibly this should be Context instead of _context. It looks to me like Context represent the actual context but its value is being cached in _context meaning that it can be uninitialized only because noone ever used it

bartonjs replied:
If the tests found a product bug, but the product bug also exists in .NET Framework, we should leave the bug for now and open a new issue to track it (so that the fix gets marked port-consider and it's obvious what part we're port-considering).

krwq replied:
@anthonylangsworth is this implementing this part correctly now? https://www.w3.org/TR/2001/REC-xml-c14n-20010315#NoNSPrefixRewriting - I'm not sure what PropagatedNamespaces should mean - I do not understand what does "namespace propagated into signature" is supposed to mean

anthonylangsworth replied:
@bartonjs @krwq I can confirm netfx also throws a `NullReferenceException`. I appear to have found and inadvertently fixed a framework bug. I'll revert this change later today. How do I raise the issue against .Net Framework?

krwq replied:
@anthonylangsworth please keep the change, please file the issue in this repo, assign to me and point to your fix. I'll take care of making sure this gets ported to full .NET. I think the test should still be expecting prefixes in the elements though - was this also like that on netfx?

bartonjs replied:
@krwq I'm pretty sure that our normal tactic here is to always have netfx-impacting changes be as small as possible; so this should stay broken for now, and get fixed as a separate issue.

@AlexGhiondea: Thoughts?

AlexGhiondea replied:
@krwq since this is a bug in Desktop, here is the process we should follow
- Port the code as-is (don't introduce a change) -- this would bring the code to par with desktop
- File a new issue for the change you are making
- Tag the issue with `netfx-port-consider` this will flag it for consideration for Desktop
- Follow the Desktop Shiproom process to get this change approved

If we start making fixes as we are porting the code there is a high chance that we are going to forget to port that change back to Desktop.

bartonjs commented on src/System.Security.Cryptography.Xml/tests/SignedXmlTest.cs:
```diff
@@ -658,7 +645,7 @@ public void DigestValue_CRLF()
             s.Position = 0;
 
-            HashAlgorithm hash = HashAlgorithm.Create("System.Security.Cryptography.SHA1CryptoServiceProvider");
+            HashAlgorithm hash = SHA1.Create();
```
HashAlgorithm : IDisposable, this should be in a using.

krwq commented:
@anthonylangsworth @tintoy @peterwurzinger let's close on this PR and create a new one.

Here are the minimum items we need to re-enable the package:
https://github.com/dotnet/corefx/issues?q=is%3Aopen+is%3Aissue+milestone%3A2.0.0+label%3ASystem.Security.Cryptography.Xml+assignee%3Akrwq

This one requires writing new tests, anything else is bugs
https://github.com/dotnet/corefx/issues/16781

after that we can start improving the package.

We're almost there! Good job everyone! If you'd like to help on any of them please comment on the issue so we do not accidentally duplicate work.

bartonjs commented on src/System.Security.Cryptography.Xml/tests/XmlDsigXsltTransformTest.cs:
```diff
@@ -182,8 +182,11 @@ public void OnlyInner()
             transform.LoadInnerXml(doc.ChildNodes);
-            Stream s = (Stream)transform.GetOutput();
-            Assert.Throws<ArgumentNullException>(() => Stream2Array(s));
+            Assert.Throws<ArgumentNullException>(() =>
```
What's supposed to ANE here?  If it's checking that transform.GetOutput() is null then assert that it's null...

krwq replied:
@bartonjs we are porting existing Mono tests right now and simply fixing errors happened during conversion from different test framework. I agree this is not ideal but I think we should do separate pass on these kind of clean ups

bartonjs replied:
@krwq I appreciate that.  But my point is that if you're making a change to the test you should make the right change.  It seems weird to me to have a two line lambda expecting an ArgumentNullException.  If the problem is that Mono and NetFx generate them from different places, add a comment.  Otherwise, make the test test what we want it to test.... whatever that may be.

krwq replied:
Converting to single line.

krwq commented on src/System.Security.Cryptography.Xml/tests/DSAKeyValueTest.cs:
```diff
@@ -88,7 +88,7 @@ public void GetXml_SameDsa()
         }
 
-        [Fact]
+        [Fact(Skip = "https://github.com/dotnet/corefx/issues/16779")]
```
@bartonjs I'm not converting to ActiveIssue while there is some active development happening and until this is fixed: https://github.com/dotnet/corefx/issues/16686 - in the issue I meant that the exactly the same test build and run against full .net framework is passing. possibly corefx test is failing when run on desktop or linux is using wrong assemblies - I'll investigate that when I get to this failure. I plan on fixing all of them before publishing the package.

krwq commented:
@anthonylangsworth I have disabled the test, left the change to the test and reverted the `?.`

krwq commented:
@bartonjs - I've changed SHA1 to be wrapped with using, changed ArgumentNullException test to have 1 line lambda and undid change introducing diff with Desktop

krwq commented:
@dotnet-bot test Innerloop Ubuntu14.04 Debug Build and Test please

krwq commented:
Thank you everyone for help here!
